healthcheck/ticker.go

```diff
@@ -1,35 +1,135 @@
+// - sends alerts of any change in health via the given channel (if non-nil) - with the new state (i.e. `true` if newly-healthy)
+// - healthchecks can be requested using requestCheckChan (sending `false` on this channel indicates that a failure has prompted this check)
+// - healthchecks will only be run at a maximum of `recoveryDuration` frequency
+func NewTickerWithAlerts(
+	duration,
+	recoveryDuration time.Duration,
+	stateChangeChan,
+	requestCheckChan chan bool,
+	clients ...Client,
+) *Ticker {
+
+	ticker := Ticker{
+		timeTicker: time.NewTicker(duration),
+		closing:    make(chan bool),
+	}
+
+	asyncCheck := make(chan bool)
+
+	// mutexChecking locks the below vars, and also locks MonitorExternal
+	var mutexChecking sync.Mutex
+	var lastCheckStarted time.Time
+	var lastCheckCompleted time.Time
+
+	// mutexCurrentHealthOK locks the below var
+	var mutexCurrentHealthOK sync.Mutex
+	var currentHealthOK bool
 
-	timeTicker := time.NewTicker(duration)
+	// create recovery ticks which initiate a healthcheck only on startup and when unhealthy
+	go func() {
+		for range time.Tick(recoveryDuration) {
+			mutexCurrentHealthOK.Lock()
+			if !currentHealthOK {
+				asyncCheck <- true
+			}
+			mutexCurrentHealthOK.Unlock()
+		}
+	}()
 
+	// main goroutine to run MonitorExternal at intervals, and send any health state changes
 	go func() {
+		for {
+			// block until closing, or a check is due (either ticker chan) or requested
+			select {
+			case <-ticker.closing:
+				return
+			case <-asyncCheck:
+			case <-ticker.timeTicker.C:
+			case claimedNewHealthOK := <-requestCheckChan:
+				mutexCurrentHealthOK.Lock()
+				logData := log.Data{"prev_health": currentHealthOK, "new_health": claimedNewHealthOK}
+				if currentHealthOK != claimedNewHealthOK {
+					if !claimedNewHealthOK {
+						// fail fast if claimedNewHealthOK is bad
+						log.Debug("health state change - fail via request", logData)
+						stateChangeChan <- claimedNewHealthOK
+						currentHealthOK = claimedNewHealthOK
+					} else {
+						// recover slowly: only change state after successful MonitorExternal()
+						log.Debug("health state change - success claimed - will check", logData)
+					}
+				} else {
+					log.Debug("healthcheck requested - no change in health", logData)
+				}
+				mutexCurrentHealthOK.Unlock()
+			}
+
+			now := time.Now()
+			mutexChecking.Lock()
+
+			// too soon for healthcheck?
+			if now.Before(lastCheckStarted.Add(recoveryDuration)) || now.Before(lastCheckCompleted.Add(recoveryDuration)) {
+				mutexChecking.Unlock()
+				log.Debug("too soon for healthcheck", log.Data{"last_check_start": lastCheckStarted, "last_check_ended": lastCheckCompleted})
 
-		for range timeTicker.C {
+			} else {
 
-			log.Debug("conducting service healthcheck", nil)
-			MonitorExternal(clients...)
+				lastCheckStarted = now
+
+				// run check in the background,  mutexChecking.Lock() applies
+				go func() {
```

Comment:
I cannot see why this go routine is required. It seems we are only allowing one healthcheck to run at a time, so can we do the check in the parent go routine?

Reply:
If `MonitorExternal` is running, and taking a minute to timeout for each external service (so can be taking several minutes), then the goroutine allows the app to send `requestCheckChan` messages to downgrade the health (fail fast) _despite this blocking_ - kind of belts-and-braces.

I suppose you could argue that there should be a separate goroutine for that listener, but this makes sense to me. Does that help? 🤔

Reply:
I see - that makes sense 👍